admin-dev/themes/new-theme/js/pages/product/price-refresher.js

```diff
@@ -0,0 +1,153 @@
+/**
+ * 2007-2020 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+import productMap from './product-map';
+import * as types from './store/mutation-types';
+
+/**
+ * Responsible for price tax excluded/included and tax rate selection inputs interaction
+ */
+export default class PriceRefresher {
+  constructor(store) {
+    this.store = store;
+    this.taxIncludedInputs = productMap.priceTaxIncludedInputs;
+    this.taxExcludedInputs = productMap.priceTaxExcludedInputs;
+    this.taxRulesGroupSelections = productMap.taxRulesGroupSelections;
+
+    this.init();
+  }
+
+  /**
+   * Initializes two way binding between DOM and state.
+   */
+  init() {
+    this.listenToDomChanges();
+    this.listenToStateChanges();
+  }
+
+  /**
+   * Listens to DOM elements changes and mutates the state accordingly.
+   */
+  listenToDomChanges() {
+    this.listenToPriceChangesInDom(this.taxIncludedInputs, true);
+    this.listenToPriceChangesInDom(this.taxExcludedInputs, false);
+    this.listenToTaxRulesGroupChangesInDom();
+  }
+
+  /**
+   * Subscribes to state mutations and updates the DOM.
+   */
+  listenToStateChanges() {
+    this.store.subscribe((mutation) => {
+      const subscribedMutationTypes = [
+        types.SET_PRICE_TAX_INCLUDED,
+        types.SET_PRICE_TAX_EXCLUDED,
+        types.SET_TAX_RULES_GROUP,
+      ];
+
+      if (!subscribedMutationTypes.includes(mutation.type)) {
+        return;
+      }
+
+      if (mutation.type === types.SET_PRICE_TAX_INCLUDED) {
+        this.updatePricesInDom(this.taxIncludedInputs, true);
+      } else if (mutation.type === types.SET_PRICE_TAX_EXCLUDED) {
+        this.updatePricesInDom(this.taxExcludedInputs, false);
+      } else {
+        this.updateTaxRulesGroupInDom();
+      }
+    });
+  }
+
+  /**
+   * Listens for price tax included/excluded inputs changes
+   *
+   * @param selector
+   * @param forTaxIncluded
+   */
+  listenToPriceChangesInDom(selector, forTaxIncluded) {
+    const elements = document.querySelectorAll(selector);
```

Comment:
I think this part could be turned into a helper class because we are gonna need this kind of behaviour a lot, something like

```js
export default class StoreBinder {
    constructor(store) {
        this.store = store;
    }

   bindInput(inputSelector, stateField, storeAction) {
        const elements = document.querySelectorAll(inputSelector);

        elements.forEach((el) => {
            el.addEventListener('change', (event) => {
                const payload = {};
                payload[stateField] = event.currentTarget.value;
                this.store.dispatch(storeAction, payload);
            });
        });
    }

   updateInput(inputSelector, stateField, mutationName) {
        this.store.subscribe((mutation) => {
            if (mutation.type === mutationName) {
                return;
            }

            const elements = document.querySelectorAll(this.taxRulesGroupSelections);
            elements.forEach((el) => {
                el.value = this.store.state[stateField];
            });
        });
    }
}
```

Reply:
Then this allows to do stuff like that:

```js
bindInputs() {
   storeBinder.bindInput(this.taxIncludedInputs, 'priceTaxIncluded', 'updatePriceTaxIncluded');
   storeBinder.updateInput(this.taxIncludedInputs, 'priceTaxIncluded', types.SET_PRICE_TAX_INCLUDED);
}
```
We could even have a two way binding helper that performs these two calls:
```js
storeBinder.twoWayBind(this.taxIncludedInputs, 'priceTaxIncluded', 'updatePriceTaxIncluded', types.SET_PRICE_TAX_INCLUDED)
```

Reply:
By the way for the `updateInput` maybe we can try and use the `watch` method from the store, this would avoid us from using the mutation type and we could use the state field name directly It would even allow enriched accessors like `'a.b.c'`, see:

https://vuex.vuejs.org/api/#watch
https://vuejs.org/v2/api/#vm-watch

Reply:
I've tried your suggestion. It seems ok with simple values bind like input.value === state.value.
But once we need something more it tears apart ( like when state contains nested object or when we need to access some data attributes from event).
Overall this "manual binding" looks like we are trying to invent the bike :(

Reply:
This is inventing the bike yes, but if we don't use vue, we're kinda forced to do this, and using vue would make it harder for now as we would need modify form_widget in order to add binding to it (v-model, v-prop..) but anyway this looks like this isn't the path we chosen.